cpp/serve/engine_actions/batch_draft.cc

```diff
@@ -0,0 +1,213 @@
+          }
+          request_ids.push_back(request->id);
+          request_internal_ids.push_back(rstate->mstates[0]->internal_id);
+          mstates.push_back(rstate->mstates[model_id]);
+          generation_cfg.push_back(request->generation_cfg);
+        }
+
+        // - Compute embeddings.
+        RECORD_EVENT(trace_recorder_, request_ids,
+                     "start propsal embedding for model_id " + std::to_string(model_id) +
+                         " draft_id " + std::to_string(draft_id));
+        NDArray embeddings =
+            models_[model_id]->TokenEmbed({IntTuple{input_tokens.begin(), input_tokens.end()}});
+        RECORD_EVENT(trace_recorder_, request_ids,
+                     "finish proposal embedding for model_id " + std::to_string(model_id) +
+                         " draft_id " + std::to_string(draft_id));
+        ICHECK_EQ(embeddings->ndim, 3);
+        ICHECK_EQ(embeddings->shape[0], 1);
+        ICHECK_EQ(embeddings->shape[1], num_requests);
+        embeddings =
+            embeddings.CreateView({num_requests, 1, embeddings->shape[2]}, embeddings->dtype);
+
+        // - Invoke model decode.
+        RECORD_EVENT(trace_recorder_, request_ids,
+                     "start proposal decode for model_id " + std::to_string(model_id) +
+                         " draft_id " + std::to_string(draft_id));
+        NDArray logits = models_[model_id]->BatchDecode(embeddings, request_internal_ids);
+        RECORD_EVENT(trace_recorder_, request_ids,
+                     "finish proposal decode for model_id " + std::to_string(model_id) +
+                         " draft_id " + std::to_string(draft_id));
+        ICHECK_EQ(logits->ndim, 3);
+        ICHECK_EQ(logits->shape[0], embeddings->shape[0]);
+        ICHECK_EQ(logits->shape[1], 1);
+
+        // - Sample tokens.
+        RECORD_EVENT(trace_recorder_, request_ids,
+                     "start proposal sampling for model_id " + std::to_string(model_id) +
+                         " draft_id " + std::to_string(draft_id));
+        std::vector<NDArray> prob_dist;
+        std::vector<float> token_probs;
+        std::vector<int32_t> next_tokens = sampler_->SampleTokens(
+            logits, models_[model_id], mstates, generation_cfg, &prob_dist, &token_probs);
+        RECORD_EVENT(trace_recorder_, request_ids,
+                     "finish proposal sampling for model_id " + std::to_string(model_id) +
+                         " draft_id " + std::to_string(draft_id));
+        ICHECK_EQ(next_tokens.size(), num_requests);
+
+        // - Update the draft tokens, prob dist, token probs of states.
+        for (int i = 0; i < num_requests; ++i) {
+          mstates[i]->draft_output_tokens.push_back(next_tokens[i]);
+          mstates[i]->draft_output_prob_dist.push_back(prob_dist[i]);
+          mstates[i]->draft_output_token_prob.push_back(token_probs[i]);
+          estate->stats.total_draft_length += 1;
+        }
```

Comment:
In order to apply frequency/presence penalty, I added a method `CommitToken` to push new tokens into the `committed_tokens` of RequestModelState https://github.com/mlc-ai/mlc-llm/blob/f13bf25c242b634cf6667f726eaf07b3ee574f16/cpp/serve/request_state.cc#L34-L37

We need a method `AddDraftToken` to add new draft tokens to RequestState. And when we roll back on the batch-verification side, we will need to decrease the occurrences in `appeared_token_ids`accordingly as well (with another method).